openproblems/tasks/chromatin_potential/methods/linear_regression.py

```diff
@@ -0,0 +1,169 @@
+import numpy as np
+import pandas as pd
+from scipy.sparse import csr_matrix
+import scanpy as sc
+import subprocess
+from ....tools.decorators import method
+
+
+def _chrom_limit(x, tss_size=2e5):
+    y = x.values
+    gene_direction = y[-1]
+    gene_start = y[-3]
+    gene_end = y[-2]
+    if gene_direction == "+":
+        return [gene_start - tss_size // 2, gene_start + tss_size // 2]
+    else:
+        return [gene_end - tss_size // 2, gene_end + tss_size // 2]
+
+
+def _get_annotation(adata):
+    from pyensembl import EnsemblRelease
+
+    # TODO: add more species
+    subprocess.call(
+        ["pyensembl", "install", "--release", "100", "--species", "mus_musculus"]
+    )
+    data = EnsemblRelease(100, species="mus_musculus")
+    # get ensemble gene coordinate
+    genes = []
+    for i in adata.var.index.map(lambda x: x.split(".")[0]):
+        try:
+            gene = data.gene_by_id(i)
+            genes.append(
+                [
+                    "chr%s" % gene.contig,
+                    gene.start,
+                    gene.end,
+                    gene.strand,
+                ]
+            )
+        except:
+            genes.append([np.nan, np.nan, np.nan, np.nan])
+    old_col = adata.var.columns.values
+    adata._var = pd.concat(
+        [adata.var, pd.DataFrame(genes, index=adata.var_names)], axis=1
+    )
+    adata._var.columns = np.hstack(
+        [old_col, np.array(["chr", "start", "end", "strand"])]
+    )
+
+
+def _atac_genes_score(adata, top_genes=500, threshold=1):
+    import pybedtools
+
+    # get annotation for TSS
+    _get_annotation(adata)
+    # basic quality control
+    sc.pp.filter_cells(adata, min_genes=200)
+    sc.pp.filter_genes(adata, min_cells=5)
+
+    adata.var["mt"] = adata.var.gene_short_name.str.startswith(
+        "mt-"
+    )  # annotate the group of mitochondrial genes as 'mt'
+
+    adata._inplace_subset_var(~pd.isnull(adata.var["mt"]))
+
+    sc.pp.calculate_qc_metrics(
+        adata, qc_vars=["mt"], percent_top=None, log1p=False, inplace=True
+    )
+
+    adata._inplace_subset_obs(adata.obs.n_genes_by_counts < 2000)
+    adata._inplace_subset_obs(adata.obs.pct_counts_mt < 10)
+
+    sc.pp.normalize_total(adata, target_sum=1e4)
+    sc.pp.log1p(adata)
+    sc.pp.highly_variable_genes(adata, n_top_genes=top_genes)
+
+    adata._inplace_subset_var(adata.var.highly_variable)
```

Comment:
Please don't use private members. Just do `adata = adata[:, adata.var.highly_variable].copy()`

Reply:
I think there is an issue if we use .copy(), .copy could not have effect in place, but private members function could have effect in place. Once used .copy(), the other values can not be inserted.

Reply:
To be concrete, actually I have tested adata = adata[:, adata.var.highly_variable].copy(), but 'gene_score' property cannot be inserted in to .obsm.

Reply:
I don't understand. 
```
>>> import numpy as np
>>> import scanpy as sc
>>> adata = sc.datasets.pbmc68k_reduced()
>>> adata
AnnData object with n_obs × n_vars = 700 × 765
    obs: 'bulk_labels', 'n_genes', 'percent_mito', 'n_counts', 'S_score', 'G2M_score', 'phase', 'louvain'
    var: 'n_counts', 'means', 'dispersions', 'dispersions_norm', 'highly_variable'
    uns: 'bulk_labels_colors', 'louvain', 'louvain_colors', 'neighbors', 'pca', 'rank_genes_groups'
    obsm: 'X_pca', 'X_umap'
    varm: 'PCs'
    obsp: 'distances', 'connectivities'
>>> sc.pp.highly_variable_genes(adata, n_top_genes=50)
>>> adata = adata[:, adata.var.highly_variable].copy()
>>> adata.obsm['gene_score'] = adata.obsm['X_pca'] * np.random.normal(0, 1, adata.shape[0])[:,None]
>>> adata
AnnData object with n_obs × n_vars = 700 × 50
    obs: 'bulk_labels', 'n_genes', 'percent_mito', 'n_counts', 'S_score', 'G2M_score', 'phase', 'louvain'
    var: 'n_counts', 'means', 'dispersions', 'dispersions_norm', 'highly_variable'
    uns: 'bulk_labels_colors', 'louvain', 'louvain_colors', 'neighbors', 'pca', 'rank_genes_groups', 'hvg'
    obsm: 'X_pca', 'X_umap', 'gene_score'
    varm: 'PCs'
    obsp: 'distances', 'connectivities'
```

Where are you having problems?

Reply:
Here is a demo if we want to wrap the operation as a function:

```
>>> import numpy as np
>>> import scanpy as sc
>>> adata = sc.datasets.pbmc68k_reduced()
>>> def change_adata(adata):
...     sc.pp.highly_variable_genes(adata, n_top_genes=50)
...     adata = adata[:, adata.var.highly_variable].copy()
...     adata.obsm['gene_score'] = adata.obsm['X_pca'] * np.random.normal(0, 1, adata.shape[0])[:,None]
...
...
>>> change_adata(adata)
/Users/qq06/anaconda3/envs/openproblem/lib/python3.8/site-packages/anndata-0.7.4-py3.8.egg/anndata/_core/anndata.py:1094: FutureWarning: is_categorical is deprecated and will be removed in a future version.  Use is_categorical_dtype instead
  if not is_categorical(df_full[k]):
>>> adata
AnnData object with n_obs × n_vars = 700 × 765
    obs: 'bulk_labels', 'n_genes', 'percent_mito', 'n_counts', 'S_score', 'G2M_score', 'phase', 'louvain'
    var: 'n_counts', 'means', 'dispersions', 'dispersions_norm', 'highly_variable'
    uns: 'bulk_labels_colors', 'louvain', 'louvain_colors', 'neighbors', 'pca', 'rank_genes_groups', 'hvg'
    obsm: 'X_pca', 'X_umap'
    varm: 'PCs'
    obsp: 'distances', 'connectivities'
>>> adata.obsm['gene_score']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/qq06/anaconda3/envs/openproblem/lib/python3.8/site-packages/anndata-0.7.4-py3.8.egg/anndata/_core/aligned_mapping.py", line 148, in __getitem__
    return self._data[key]
KeyError: 'gene_score'
```

Reply:
Why do we want to wrap it as a function? Of course it won't change it in-place.

Reply:
Because this is the way your previous task shows, I just follow them to do so.

Reply:
If you modify not in-place inside a function, you should `return adata` and do `adata = change_adata(adata)` to keep youyr changes rather than discarding them.

Reply:
However you choose to do this (either by directly calling `adata = adata[:, adata.var.highly_variable].copy()` or by putting that inside a function) we should not be using private members of `anndata` as they are liable to change / break.

Reply:
Got it. Please see the latest commit.